lib/errorResponseInterceptor.ts

```diff
@@ -53,3 +56,52 @@ export function errorResponseInterceptor(axiosError) {
 
   return Promise.reject(axiosError);
 }
+
+/**
+ * Axios interceptor to convert errors for streaming cases into
+ * JSON objects.
+ *
+ * This means users can handle rejections for AsStream cases in
+ * the same way as normal error rejections and we can apply the
+ * errorResponseInterceptor to augment the errors.
+ *
+ * @param axiosError
+ * @returns rejected promise with the stream body transoformed
+ */
+export async function errorResponseStreamConverter(axiosError) {
+  if (
+    axiosError.response && // must have a response
+    axiosError.response.status >= 400 &&
+    // must have data:
+    axiosError.response.data &&
+    // Must be a JSON response
+    axiosError.response?.headers['content-type']?.startsWith(
+      'application/json'
+    ) &&
+    // must not already be deserialized by axios
+    axiosError.config.responseType !== 'json' &&
+    // must be a stream
+    axiosError.response.data instanceof Readable
+  ) {
+    let data = '';
+    await pipeline(
+      axiosError.response.data,
+      new Writable({
+        write: (c, e, cb) => {
+          data += c;
+          cb();
+        },
+      })
+    );
+    try {
+      axiosError.response.data = JSON.parse(data);
+      axiosError.config.responseType = 'json';
```

Comment:
Just to confirm - will this make axios to set content type and content length correctly?

Reply:
good question, will double check, but neither `content-type` nor `content-length` should change (unless we had some weird encoding switch). I'm not super comfortable with setting this toggle, but I did it to make sure the converted stream object is detected by the error augmentation interceptor (which uses `responseType`) - perhaps there is a better way.

Reply:
So the default value here is `json` and it is part of the [request config](https://github.com/axios/axios?tab=readme-ov-file#request-config). The `*AsStream` methods actually set it to value `stream` instead.
It was sloppy for me to edit this as a hack to get the other interceptor in play - I don't think we should modify the request config after the request is made.

Possible options:
i. call the other interceptor directly from this new interceptor (since otherwise it will bypassed if the `responseType` remains `json`.
ii. change the check in the other interceptor to something other than `responseType` (like maybe check it isn't a `Readable`, although I'm not sure that covers all the possibilities)
iii. set some other flag on the `axiosError` to allow the other interceptor to detect these converted errors with something like `responseType === 'json' || (responseType === 'stream' and someFlag)`

or maybe something else? Thoughts please.

Reply:
Option `iii` in eec09662 for consideration

Reply:
As per https://github.com/IBM/cloudant-node-sdk/pull/1663#discussion_r1828868040 - we could consider just checking `content-type`, though I think it is a little risky. Maybe content-type + responseType = json or stream?

Reply:
I have no strong feelings, we can have https://github.com/IBM/cloudant-node-sdk/commit/eec09662cb8e0ef8718741ef61448f75870ce32d as-is.

Reply:
Well, if we are relying on interceptors order, i.e. transformer running first and setting `someFlag`, we can as well skip the flag, trust it did its job and just check for `content-type & (json || stream)`, but we'll be paying price of double failure to deserialize json if it is broken.

In this perspective option 1 feels as a most practical one. This makes me to think if we really want this to be a breaking change. As it is right now it is not that different from golang, where `*AsStream` returns errors as `ReaderCloser`, which we are reading, augmenting and re-wrapping.  So maybe we should do the same here - on `stream` do what this interceptor's doing and then return the new error as `Readable`.

Reply:
This stream/object conversion should be considered independently of the error augmentation, but of course they need to work in tandem.

None of the other lang SDKs force a user to read a stream to get an error object from an exception, it is done for them and I think the same should happen in Node. Arguably it's a core bug that it doesn't convert the stream in error cases (we could try for a PR there instead, but we might want this interceptor as a bridging solution as I don't know how long it might take especially since it would be breaking).

I'm ok to do option 1, and the error augmentation interceptor will just skip the output of this interceptor because it is already augmented, but I'd still return an object and not a `Readable` (because the core doesn't handle the `Readable` unlike in Go).

Reply:
Ok, fair enough, I had to read core's code, but you right, in go's `Request` core's reading returned body in `processErrorResponse` and converts it into an error, so that's different from how node's dong this. 

I agree that stream conversion should be considered independently of the error augmentation and honestly for option one I thought other way around - for error aug to do streaming conversion, not for this stream conversion to do aug error. My concern is that for streaming broken JSON we are wasting cycles on de-serialization attempt twice and then returning neither original response not augmented response.

Maybe the right thing to do is not to do JSON convertion here, but read stream into string and pass it along? Then in error aug interceptor we don't need a special flag and can just trust that if it is stream and error we did the reading already and can use `.data` same as not streaming response?

Reply:
>My concern is that for streaming broken JSON we are wasting cycles on de-serialization attempt twice and then returning neither original response not augmented response.

That's a fair concern, but in the `json` case the parse is done by `axios` before we (or the core) ever see `data`. So in a broken `json` case axios will have set a `string` for `data`. In the `stream` case the stream converter will try to deserialize and fail and set data as `string` which, I believe, matches the `axios` behaviour.

So I think in either case we have only 1 attempt at parsing the JSON - the `errorResponseInterceptor` doesn't ever attempt to parse JSON because in Node `json` case `.data` is already an `object` (because axios did the parse before we ever saw it). In broken json cases `data` would be a string and fail the property checks and do nothing, falling back to the core's `string` behaviour.

>Maybe the right thing to do is not to do JSON convertion here, but read stream into string and pass it along? Then in error aug interceptor we don't need a special flag and can just trust that if it is stream and error we did the reading already and can use `.data` same as not streaming response?

I see what you're getting at and it sounds reasonable, but for non-broken JSON cases we never have a `string` in `data` we have either a `Readable` or an `object`. I'll try to just use those checks so we can avoid the flag.